mesonpy/_editable.py

```diff
@@ -2,183 +2,298 @@
+                    for entry in walk(src):
+                        tree[(*path.parts[1:], *entry.parts)] = os.path.join(src, *entry.parts)
+                else:
+                    tree[path.parts[1:]] = src
+    return tree
+
+
+class MesonpyMetaFinder(importlib.abc.MetaPathFinder):
+    def __init__(self, names: Set[str], path: str, args: List[str], verbose: bool):
+        self._top_level_modules = names
+        self._build_path = path
+        self._compile_args = args
+        self._verbose = verbose
+        self._loaders: List[Tuple[type, str]] = []
+        for loader, suffixes in LOADERS:
+            self._loaders.extend((loader, suffix) for suffix in suffixes)
 
     def find_spec(
-        self,
-        fullname: str,
-        path: Optional[Sequence[Union[str, bytes]]],
-        target: Optional[ModuleType] = None,
-    ) -> None:
-        # if it's one of our modules, trigger a rebuild
+            self,
+            fullname: str,
+            path: Optional[Sequence[Union[bytes, str]]] = None,
+            target: Optional[ModuleType] = None
+    ) -> Optional[importlib.machinery.ModuleSpec]:
         if fullname.split('.', maxsplit=1)[0] in self._top_level_modules:
-            self.rebuild()
-            # prepend the project path to sys.path, so that the normal finder
-            # can find our modules
-            # we prepend so that our path comes before the current path (if
-            # the interpreter is run with -m), see gh-239
-            if sys.path[:len(self._import_paths)] != self._import_paths:
-                for path in self._import_paths:
-                    if path in sys.path:
-                        sys.path.remove(path)
-                sys.path = self._import_paths + sys.path
-        # return none (meaning we "didn't find" the module) and let the normal
-        # finders find/import it
+            if self._build_path in os.environ.get(MARKER, '').split(os.pathsep):
+                return None
+            namespace = False
+            tree = self.rebuild()
+            parts = fullname.split('.')
+            # look for a package
+            package = tree.get(tuple(parts))
+            if isinstance(package, Node):
+                for loader, suffix in self._loaders:
+                    src = package.get('__init__' + suffix)
+                    if isinstance(src, str):
+                        return build_module_spec(loader, fullname, src, package)
+                else:
+                    namespace = True
+            # look for a module
+            for loader, suffix in self._loaders:
+                src = tree.get((*parts[:-1], parts[-1] + suffix))
+                if isinstance(src, str):
+                    return build_module_spec(loader, fullname, src, None)
+            # namespace
+            if namespace:
+                spec = importlib.machinery.ModuleSpec(fullname, None)
+                spec.submodule_search_locations = []
+                return spec
```

Comment:
Makes it slightly easier to read for me

```suggestion
            if self._build_path in os.environ.get(MARKER, '').split(os.pathsep):
                return None
            namespace = False
            tree = self.rebuild()
            parts = fullname.split('.')

            # look for a package
            package = tree.get(tuple(parts))
            if isinstance(package, Node):
                for loader, suffix in self._loaders:
                    src = package.get('__init__' + suffix)
                    if isinstance(src, str):
                        return build_module_spec(loader, fullname, src, package)
                else:
                    namespace = True

            # look for a module
            for loader, suffix in self._loaders:
                src = tree.get((*parts[:-1], parts[-1] + suffix))
                if isinstance(src, str):
                    return build_module_spec(loader, fullname, src, None)

            # namespace
            if namespace:
                spec = importlib.machinery.ModuleSpec(fullname, None)
                spec.submodule_search_locations = []
                return spec

```